src/pyop/storage.py

```diff
@@ -56,14 +154,17 @@ def __init__(self, db_uri, db_name=None,
         if db_uri is None:
             raise ValueError('db_uri not supplied')
 
-        self._db_uri = db_uri
-        self._database_name = db_name
         self._sanitized_uri = None
 
         self._parsed_uri = pymongo.uri_parser.parse_uri(db_uri)
 
         if self._parsed_uri.get('database') is None:
+            if db_name is None:
+                raise ValueError(
+                    "Database name must be provided either in the URI or as an argument"
+                )
             self._parsed_uri['database'] = db_name
+        self._database_name = self._parsed_uri['database']
```

Comment:
I can see this being a bit simpler than nested conditions.

```suggestion
        db_name = self._parsed_uri.get('database') or db_name
        if db_name is None:
            raise ValueError(
                "Database name must be provided either in the URI or as an argument"
            )
        self._database_name = self._parsed_uri['database'] = db_name
```

Reply:
Thanks, updated